doc/specs/#674 - PowerShell cmdlets.md

```diff
@@ -0,0 +1,434 @@
+
+@ThomasNieto was the first to create an Issue for [Native PowerShell Support](https://github.com/microsoft/winget-cli/issues/221).
+This is one of the [most desired features](https://github.com/microsoft/winget-cli/issues?page=1&q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3AIssue-Feature) for the Windows Package Manager.
+
+## Solution Design
+
+The initial solution will start with [PowerShell Crescendo Preview.3](https://devblogs.microsoft.com/powershell/announcing-powershell-crescendo-preview-3/).
+Many of the major considerations are related to proper PowerShell nouns and verbs. @MilkFather suggested [command aliases](https://github.com/microsoft/winget-cli/issues/380) for the winget.exe
+client and @Jaifroid also suggested [find](https://github.com/microsoft/winget-cli/issues/1299) rather than search.
+
+## UI/UX Design
+
+The main UI/UX differences will be in the command names, parameters, and how the output is formatted for users. For the sake of
+consistency the sub headings below will map to existing commands in winget.exe to begin drafting some syntax proposals. In order
+to keep the discussion focussed, I'll create GitHub discussions for each of these and link them here as we work through driving
+consensus.
+
+The PowerShell module name will be "WindowsPackageManager"
+The Adjective "WinGet" may change, but is used consistently in the cmdlets until such time as we agree on the proper term.
+
+>Note: Naming things is hard, and one of the conventions of PowerShell is to make nouns singular.
+---
+
+### Client command `winget install`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Install-WinGetPackage
+```
+**Parameters**
+
+-Manifest
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Version
+
+-Source
+
+-Scope
+
+-Exact
+
+-Interactive
+
+-Silent
+>Note: we may switch the default to "silent" from "silent with progress" in winget.exe so there may be changes here.
+
+-Locale
+
+-Log
+
+-Override
+
+-Location
+
+---
+
+### Client command `winget show`
+
+**PowerShell Cmdlet**
+```PowerShell
+Get-WinGetPackageVersion
+```
+
+-Detail
+
+-All
+
+-Version
+
+**Examples**
+
+>Note: The current behavior for `winget show vscode` would look like the example below.
+
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -Detail
+```
+
+>Note: The current behavior for `winget show vscode --versions` would look like the example below.
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -All
+```
+
+---
+
+### Client command `winget source`
+
+**PowerShell Cmdlets**
+
+```PowerShell
+Add-WinGetSource
+```
+
+```PowerShell
+Remove-WinGetSource
+```
+
+```PowerShell
+Get-WinGetSource
+```
+
+```PowerShell
+Reset-WinGetSource
+```
+
+**Example**
+> Note: `winget source export` was added as a helper for adding approved sources to Group Policy. To get the same values with PowerShell the example blow would perform the same behavior.
+
+
+```PowerShell
+Get-WinGetSource | convertto-json -compress
```

Comment:
Since this would entirely rely on the output object structure this approach could break easily. It might make more sense to handle this via a `-Raw` parameter, e.g.
```powershell
Get-WinGetSource -Raw
```

Reply:
The cmdlets shoudl emit either .NET Objects (for consumption by other PowerShell cmdlets along with JSON where necessary.

Reply:
NOte - if the intention is to handle cmdlets using Crescendo - then this can be met, _provided_ the team do a good job at writing the necessary output handelers. And for the record, with Cresendo, this is NOT an easy task unless you are outstandingly good with regular expressions.

With that said, the cmdlets must emit objects - and some of those objects can contain json where json is relevant.